dateparser/search/search.py

```diff
@@ -170,25 +170,32 @@ def __init__(self):
         self.loader = LocaleDataLoader()
         self.available_language_map = self.loader.get_locale_map()
         self.search = _ExactLanguageSearch(self.loader)
-
-    def detect_language(self, text, languages):
-        if isinstance(languages, (list, tuple, Set)):
-
-            if all([language in self.available_language_map for language in languages]):
-                languages = [self.available_language_map[language] for language in languages]
-            else:
-                unsupported_languages = set(languages) - set(self.available_language_map.keys())
-                raise ValueError(
-                    "Unknown language(s): %s" % ', '.join(map(repr, unsupported_languages)))
-        elif languages is not None:
-            raise TypeError("languages argument must be a list (%r given)" % type(languages))
-
-        if languages:
-            self.language_detector = FullTextLanguageDetector(languages=languages)
+    
+    @apply_settings
+    def detect_language(self, text, languages, settings=None):
+        language = str
+        
+        if settings.LANGUAGE_DETECTION_ENABLED:
+            from dateparser.custom_lang_detect import detect_languages
+            detect_languages = detect_languages(settings=settings)
+            language = detect_languages(text)[0]
         else:
-            self.language_detector = FullTextLanguageDetector(list(self.available_language_map.values()))
-
-        return self.language_detector._best_language(text)
+            if isinstance(languages, (list, tuple, Set)):
+                if all([language in self.available_language_map for language in languages]):
+                    languages = [self.available_language_map[language] for language in languages]
+                else:
+                    unsupported_languages = set(languages) - set(self.available_language_map.keys())
+                    raise ValueError(
+                        "Unknown language(s): %s" % ', '.join(map(repr, unsupported_languages)))
+            elif languages is not None:
+                raise TypeError("languages argument must be a list (%r given)" % type(languages))
+            
+            if languages:
+                self.language_detector = FullTextLanguageDetector(languages=languages)
+            else:
+                self.language_detector = FullTextLanguageDetector(list(self.available_language_map.values()))     
+            language = self.language_detector._best_language(text)
+        return language
```

Comment:
These new changes are creating a complex nesting with the if-else statements. To fix this you should use returns:

```python
if settings.LANGUAGE_DETECTION_ENABLED:
    ...
    return detect_languages(text)[0]

if isinstance(languages, (list, tuple, Set)):
    ...
    return self.language_detector._best_language(text)
```

(This is, put it how it was before but adding your `if` and `return` before.)